File: src/main/java/dev/haarolean/gnutardbot/abilities/ReportMessageAbility.java
package dev.haarolean.gnutardbot.abilities;

import dev.haarolean.gnutardbot.TardBot;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;
import org.telegram.abilitybots.api.objects.Ability;
import org.telegram.abilitybots.api.objects.Locality;
import org.telegram.abilitybots.api.objects.MessageContext;
import org.telegram.abilitybots.api.objects.Privacy;
import org.telegram.telegrambots.meta.api.methods.ForwardMessage;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.util.Objects;

import static org.springframework.beans.factory.config.BeanDefinition.SCOPE_PROTOTYPE;

@Component
@Scope(SCOPE_PROTOTYPE)
public class ReportMessageAbility implements AbilityProvider {

    private final TardBot bot;

    public ReportMessageAbility(TardBot bot) {
        this.bot = bot;
    }

    @Override
    public Ability buildAbility() {
        return Ability
                .builder()
                .name("report")
                .locality(Locality.GROUP)
                .privacy(Privacy.PUBLIC)
                .action(this::reportMessage)
                .post(bot::deleteMessage)
                .build();
    }

    private void reportMessage(MessageContext ctx) {
        var message = ctx.update().getMessage();
        var sender = message.getFrom();
        long senderId = sender.getId();
        var chatId = message.getChatId().toString();

        if (!message.isReply()) return;

        var reply = message.getReplyToMessage();
        long reportedUserId = reply.getFrom().getId();

        if(!Objects.equals(message.getChatId(), reply.getChatId())) return;
        if (bot.isAdmin(reportedUserId)) return;
        if (bot.isGroupAdmin(ctx.update(), reportedUserId)) return;

        bot.admins().forEach(adminId -> {
            try {
                bot.execute(new ForwardMessage(adminId.toString(), chatId, message.getMessageId()));
                bot.execute(new ForwardMessage(adminId.toString(), chatId, reply.getMessageId()));
            } catch (TelegramApiException e) {
                if(e.getMessage().contains("[400]") || e.getMessage().contains("[403]")) return;
                throw new RuntimeException(e);
            }
        });
    }


}

File: src/main/kotlin/dev/haarolean/gnutardbot/abilities/ReportMessageAbility.kt
package dev.haarolean.gnutardbot.abilities

import dev.haarolean.gnutardbot.TardBot
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import org.telegram.abilitybots.api.objects.Ability
import org.telegram.abilitybots.api.objects.Locality
import org.telegram.abilitybots.api.objects.MessageContext
import org.telegram.abilitybots.api.objects.Privacy
import org.telegram.telegrambots.meta.api.methods.ForwardMessage
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class ReportMessageAbility(private val bot: TardBot) : AbilityProvider {
    override fun buildAbility(): Ability {
        return Ability
            .builder()
            .name("report")
            .locality(Locality.GROUP)
            .privacy(Privacy.PUBLIC)
            .action { reportMessage(it) }
            .post { bot.deleteMessage(it) }
            .build()
    }

    private fun reportMessage(ctx: MessageContext) {
        val message = ctx.update().message
        val chatId = message.chatId.toString()
        if (!message.isReply) return
        val reply = message.replyToMessage
        val reportedUserId = reply.from.id
        if (message.chatId != reply.chatId) return
        if (bot.isAdmin(reportedUserId)) return
        if (bot.isGroupAdmin(ctx.update(), reportedUserId)) return
        bot.admins().forEach {
            try {
                bot.execute(ForwardMessage(it.toString(), chatId, message.messageId))
                bot.execute(ForwardMessage(it.toString(), chatId, reply.messageId))
            } catch (e: TelegramApiException) {
                if (e.message?.contains("[400]") == true
                        || e.message?.contains("[403]") == true) return
                throw RuntimeException(e)
            }
        }
    }
}
